enterprise/server/remote_execution/containers/ociruntime/ociruntime.go

```diff
@@ -852,35 +857,91 @@ func (c *ociContainer) invokeRuntime(ctx context.Context, command *repb.Command,
 	return result
 }
 
-func statDevice(path string) (*specs.LinuxDevice, error) {
-	info, err := os.Stat(path)
+func getUser(ctx context.Context, image *Image, rootfsPath string, dockerUserProp string, dockerForceRootProp bool) (*specs.User, error) {
+	// TODO: for rootless support we'll need to handle the case where the
+	// executor user doesn't have permissions to access files created as the
+	// requested user ID
+	spec := ""
+	if image != nil {
+		spec = image.Config.User
+	}
+	if dockerUserProp != "" {
+		spec = dockerUserProp
+	}
+	if dockerForceRootProp {
+		spec = "0"
+	}
+	if spec == "" {
+		// Inherit the current uid/gid.
+		spec = fmt.Sprintf("%d:%d", os.Getuid(), os.Getgid())
+	}
+
+	user, group, err := container.ParseUserGroup(spec)
 	if err != nil {
-		return nil, fmt.Errorf("could not stat device file: %v", err)
+		return nil, fmt.Errorf(`invalid "USER[:GROUP]" spec %q`, spec)
+	}
+
+	var uid, gid int
+	username := user.Name
+
+	// If the user is non-numeric then we need to look it up from /etc/passwd.
+	// If no gid is specified then we need to find the user entry in /etc/passwd
+	// to know what group they are in.
+	if user.Name != "" || group == nil {
+		userRecord, err := unixcred.LookupUser(filepath.Join(rootfsPath, "/etc/passwd"), user)
+		if (err == unixcred.ErrUserNotFound || os.IsNotExist(err)) && user.Name == "" {
+			// If no user was found in /etc/passwd and we specified only a
+			// numeric user ID then just set the group ID to 0 (root).
```

Comment:
Why is it safe for us to set group id to root in this case?

Reply:
The group ID here is being set within the container and does not actually grant true root access because of various container isolation features (seccomp, capabilities, etc.)

As to why we set root and not some other group ID, I think it's because gid 0 is probably the only gid that is guaranteed to exist. I added this to the comment.